bpo-35537: use os.posix_spawn in subprocess

nanjekyejoannah opened the pull request:
This patch uses os.posix_spawn in subprocess.

posix_spawn() is faster but it's also safer: it allows us to do a lot of "actions" before exec(), before executing the new program. For example, you can close files and control signals. 

@vstinner this is the patch without any options.


<!-- issue-number: [bpo-35537](https://bugs.python.org/issue35537) -->
https://bugs.python.org/issue35537
<!-- /issue-number -->

nanjekyejoannah committed:
use os.posix_spawn in subprocess

vstinner commented on Lib/subprocess.py:
```diff
@@ -1414,6 +1419,16 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
+                    and not pass_fds
+                    and cwd is None
+                    and env is None
+                    and restore_signals is False
```
"and not restore_signals" would be better. I prefer to avoid "is" to compare with True/False.

vstinner commented on Lib/subprocess.py:
```diff
@@ -1390,6 +1390,11 @@ def _get_handles(self, stdin, stdout, stderr):
 
+        def _posix_spawn(self, args, executable):
+            """Execute program using os.posix_spawn()."""
+            env = dict(os.environ)
```
Is it really needeed to copy os.environ?

vstinner replied:
I suggest to pass directly os.environ to os.posix_spawn().

vstinner commented on Misc/NEWS.d/next/Library/2018-12-19-17-20-03.bpo-35537.z4E7aA.rst:
```diff
@@ -0,0 +1 @@
+subprocess.Popen can now use posix_spawn() in some cases.
```
For better rendering (create links), I suggest:

```suggestion
:class:`subprocess.Popen` can now use :func:`os.posix_spawn` in some cases.
```

vstinner reviewed (approved):
LGTM, but I will wait until minor remarks are addressed.

izbyshev commented on Lib/subprocess.py:
```diff
@@ -1414,6 +1419,16 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
+                    and cwd is None
+                    and env is None
+                    and restore_signals is False
+                    and not start_new_session):
```
The condition is too weak. It doesn't check for redirections at all, so e.g. `stdout=subprocess.PIPE` won't work: it will leave both pipe ends open in the parent, but the child won't inherit anything because both ends have `O_CLOEXEC`.

Another problem is that `os.posix_spawn` never looks in `PATH`, so e.g. `subprocess.call(['sleep', '10'], close_fds=False, restore_signals=False)` doesn't work with this patch. 

Looks like some tests that would catch the above are needed :)

vstinner replied:
Right. I suggest to add more tests:

```
and os.path.isabs(executable)
and self.stdin is self.stdout is self.stderr is None)
```

vstinner replied:
> Looks like some tests that would catch the above are needed :)

test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

@nanjekyejoannah is my mentoree and we are working together to slowly use posix_spawn() in more and more cases. I prefer to work step by step, and so support more and more options, but start from the bare minimum, and check at each step that everything is fine.

izbyshev replied:
`os.path.isabs(executable)` is a bit strong. Using `os.sep not in executable` is enough: it's exactly what `execvp` and `posix_spawnp` do when they decide whether to do `PATH` lookup. But IMO even better would be to add `os.posix_spawnp` and use it instead.

> test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

It's a bit scary for me that basic functionality broke (albeit under these rare conditions) and no test failed. Basically, it indicates that if users pass `restore_signals=False`, they're on their own (`close_fds` seems to be better tested).

vstinner replied:
Don't worry, os.posix_spawn() support will be much more complete before Python 3.8 final. I'm working with @nanjekyejoannah to support restore_signals=True for example ;-)

vstinner replied:
> But IMO even better would be to add os.posix_spawnp and use it instead.

Aha, I didn't know posix_spawnp(), that's interesting. It is available on macOS and FreeBSD.

Again, I prefer to keep the first implementation as simple as possible. os.posix_spawnp() can be added later.

> os.path.isabs(executable) is a bit strong. Using os.sep not in executable is enough: it's exactly what execvp and posix_spawnp do when they decide whether to do PATH lookup.

shutil.which() uses os.path.dirname(). So I suggest to check "and os.path.dirname(executable)" instead of "and os.path.isabs(executable)".

vstinner replied:
Update, according to the discussion, the following check would be better:

```
and os.path.dirname(executable)
and self.stdin is self.stdout is self.stderr is None)
```

vstinner commented on Lib/subprocess.py:
```diff
@@ -1414,6 +1419,16 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
+                    and env is None
+                    and restore_signals is False
+                    and not start_new_session):
+                return self._posix_spawn(args, executable)
```
I suggest to pass (executable, args) to follow os.posix_spawn() API. The function has not return value, so I suggest to write instead:

```suggestion
                 self._posix_spawn(executable, args)
                 return
```

vstinner reviewed (commented):
Can you please add a sentence to the Optimization section of Doc/whatsnew/3.8.rst? Something like:

"subprocess.Popen is now up to 60x faster on Linux and 106x faster on FreeBSD when os.posix_spawn() can be used."

nanjekyejoannah committed:
use os.posix_spawn in subprocess

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Lib/subprocess.py:
```diff
@@ -1390,6 +1390,11 @@ def _get_handles(self, stdin, stdout, stderr):
 
+        def _posix_spawn(self, args, executable):
+            """Execute program using os.posix_spawn()."""
+            env = os.environ
```
nitpick: this variable is useless, just pass directly os.environ to posix_spawn().

vstinner reviewed (approved):
LGTM.

izbyshev commented on Lib/subprocess.py:
```diff
@@ -1414,6 +1419,19 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
+                    and not start_new_session
+                    and os.path.dirname(executable)
+                    and self.stdin is self.stdout is self.stderr is None):
+                self._posix_spawn(args, executable)
```
Checking `self.std*` is still not enough. Those fields are for the parent's ends of pipes, but if a caller redirects to a descriptor, there is no such end. For example, `subprocess.call(['/bin/sleep', '1000'], stdout=fd, close_fds=False, restore_signals=False)` doesn't work properly.

vstinner replied:
@izbyshev: Oh right! I fixed the test:  test  (p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite) instead of (self.stdin, self.stdout, self.stderr). Is it better now?

izbyshev replied:
Yes, LGTM now.

There is another potential problem with `posix_spawn`: [asynchronous error notification](http://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn.html). `_posixsubprocess` uses a pipe to notify the parent about all errors up to and including `execve` errors, so its style is synchronous. Glibc and musl use the pipe-based notification too, but they are not required to do that, and I don't know what other implementations do.

vstinner committed:
Fix test deciding if posix_spawn() can be used

Test  (p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite)
instead of (self.stdin, self.stdout, self.stderr).
